File: mcp-app/src/lib/ig-catalog.ts
import type { IgCatalog, IgDef, Preset } from "../types"
import catalog from "../../../shared/ig/catalog.json"

// Read-only IG catalog, single source of truth shared with the backend
// (shared/ig/catalog.json → backend/core/ig_catalog.py). Drives the config UI
// (which resource types and coding systems to render); presets store only deltas
// over these defaults. Adding an IG is a JSON edit picked up by both sides.

export const IG_CATALOG = catalog as unknown as IgCatalog

export const BASE_IG = IG_CATALOG.base[0].id

export function igById(id: string | null | undefined): IgDef | undefined {
  if (!id) return undefined
  return [...IG_CATALOG.base, ...IG_CATALOG.specialties].find((d) => d.id === id)
}

export interface FixedGroup {
  title: string
  codes: { system: string; code: string; display: string }[]
}

export interface ProfileGroup {
  title: string
  profiles: string[]
}

// Effective stack for a resource type: base, then specialty overlay.
function stackFor(preset: Preset) {
  const base = igById(preset.ig.base) ?? IG_CATALOG.base[0]
  const specialty = preset.ig.specialty ? igById(preset.ig.specialty) : undefined
  return { base, specialty }
}

// FHIR profiles the active stack applies to a resource, grouped by IG — the
// "what's in the profile" surface in the config UI (read-only).
export function profileGroupsFor(preset: Preset, rt: string): ProfileGroup[] {
  const { base, specialty } = stackFor(preset)
  const groups: ProfileGroup[] = []
  const baseP = base.resources[rt]?.profiles
  if (baseP?.length) groups.push({ title: base.title, profiles: baseP })
  const specP = specialty?.resources[rt]?.profiles
  if (specP?.length) groups.push({ title: specialty!.title, profiles: specP })
  return groups
}

// Coding systems bound for a resource — specialty overlay wins, else base.
export function codingSystemsFor(preset: Preset, rt: string): string[] {
  const { base, specialty } = stackFor(preset)
  return specialty?.resources[rt]?.coding.systems ?? base.resources[rt]?.coding.systems ?? []
}

// Effective inclusion (specialty overlay wins).
export function inclusionFor(preset: Preset, rt: string): "required" | "supported" | "optional" {
  const { base, specialty } = stackFor(preset)
  return specialty?.resources[rt]?.inclusion ?? base.resources[rt]?.inclusion ?? "optional"
}

// "…/StructureDefinition/mcode-primary-cancer-condition" → "primary cancer condition"
export function profileLabel(url: string): string {
  const slug = url.split("/").pop() ?? url
  return slug.replace(/^(mcode|us-core)-/, "").replace(/-/g, " ")
}

// Codes the active profiles pin for a resource type, grouped by IG (base, then
// specialty) — read-only in the config UI. Mirrors backend ig_catalog.fixed_codings.
export function fixedGroupsFor(preset: Preset, rt: string): FixedGroup[] {
  const groups: FixedGroup[] = []
  const base = igById(preset.ig.base) ?? IG_CATALOG.base[0]
  const baseFixed = base.resources[rt]?.fixed
  if (baseFixed?.length) groups.push({ title: base.title, codes: baseFixed })
  const specialty = preset.ig.specialty ? igById(preset.ig.specialty) : undefined
  const specFixed = specialty?.resources[rt]?.fixed
  if (specFixed?.length) groups.push({ title: specialty!.title, codes: specFixed })
  return groups
}

export function emptyPreset(id: string, name: string): Preset {
  return { id, name, ig: { base: BASE_IG, specialty: null }, resources: {}, coding: {}, prompts: {}, extensions: [] }
}

File: backend/core/ig_catalog.py
"""Backend reader for the shared IG catalog.

Single source of truth with the frontend: `shared/ig/catalog.json` drives both
the config UI and the backend's specialty -> profile resolution. Adding an IG is
a JSON edit, picked up by both sides. The catalog stays declarative; selecting
which profile actually applies to a given resource is the builder's job (see
`candidate profiles` below).
"""
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

from core.systems import SYSTEM_URIS

_CATALOG_PATH = Path(__file__).resolve().parents[2] / "shared" / "ig" / "catalog.json"


@lru_cache(maxsize=1)
def _catalog() -> dict:
    return json.loads(_CATALOG_PATH.read_text(encoding="utf-8"))


def _ig_by_id(ig_id: str) -> dict | None:
    cat = _catalog()
    for d in (*cat.get("base", []), *cat.get("specialties", [])):
        if isinstance(d, dict) and d.get("id") == ig_id:
            return d
    return None


def specialty_candidate_profiles(specialty_id: str | None) -> dict[str, list[str]]:
    """Per-resource-type candidate profiles a specialty IG contributes.

    These are *candidates*, not unconditional overlays: a resource is tagged with
    at most one (e.g. a Condition is primary OR secondary cancer), so a downstream
    builder/classifier selects the fitting profile. Empty for an unknown/None IG.
    """
    ig = _ig_by_id(specialty_id) if specialty_id else None
    if not ig:
        return {}
    out: dict[str, list[str]] = {}
    for rt, rule in (ig.get("resources") or {}).items():
        profiles = [p for p in (rule.get("profiles") or []) if isinstance(p, str)]
        if profiles:
            out[rt] = profiles
    return out


def fixed_codings(base_id: str | None, specialty_id: str | None, rt: str) -> list[dict]:
    """Codes a resource type's active profiles pin (declared per IG in the catalog).

    The pipeline assigns these deterministically, bypassing retrieval; surfaced
    read-only in the config UI and always permitted by the codeset allow-list so a
    restrictive preset can't drop a conformance-required fixed code. `system` is the
    canonical URI; entries from base + specialty are merged (deduped by system+code).
    """
    out: list[dict] = []
    seen: set[tuple[str, str]] = set()
    for ig_id in (base_id, specialty_id):
        ig = _ig_by_id(ig_id) if ig_id else None
        if not ig:
            continue
        for e in ((ig.get("resources") or {}).get(rt) or {}).get("fixed") or []:
            code = e.get("code")
            uri = SYSTEM_URIS.get(e.get("system"), e.get("system"))
            if not code or not uri or (uri, code) in seen:
                continue
            seen.add((uri, code))
            out.append({"system": uri, "code": code, "display": e.get("display", "")})
    return out
